### ISSUE TITLE
Should P4_14 and/or P4_16 implement `parser_value_set mask 0x0ff0` construct?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#2188]
@antoninbas @vgurevich 

While adding more full support for parser value sets in P4_16 to p4c and behavioral-model code, Antonin Bas pointed out that there is code in behavioral-model intended to support code that looks like this in P4_14, which p4c accepts today without any errors:
```
parser_value_set pvs0;

parser start {
    return parse_ethernet;
}

parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x0ff0 : ingress;       //  <--- this is the most significant line of the example
        default : ingress;
    }
}
```
Today p4c when invoked to generate a BMv2 JSON file for the v1model architecture, or when used to auto-translate P4_14 source code to P4_16 source code, produces exactly the same output regardless of whether the line marked in the example above is as shown, or when you remove the `mask 0x0ff0` part.

The P4_14 language specification does not mention this possibility, but p4c seems to implement it anyway, or at least silently accept it and at least somewhere in the implementation, either ignore it, or pay attention to it for a while but forget about it before issuing BMv2 JSON or P4_16 program output.

Questions:

+ Should the output be different for these two programs?
+ Should p4c treat the program with `mask 0x0ff0` present as a syntax error?
+ What is the intended difference in behavior?
+ If the behavior should be different, how should it be represented differently when auto-translating to P4_16 source code?

Note that v1.0.5 of the P4_14 language spec does not mention this syntax at all.  In section 4.4 "BNF" v1.0.5 of the P4_14 language spec gives this grammar rule:
```
value_or_masked ::=
    field_value | field_value mask field_value | value_set_name
```
Note that there is no `value_set_name mask field_value` on the right hand side mentioned.

Commands I used to compare outputs.  pvs1.p4 and pvs2.p4 are in the attached zip file.
```
$ p4test --std p4-14 pvs1.p4 --pp pvs1b.p4
$ p4test --std p4-14 pvs2.p4 --pp pvs2b.p4
$ diff pv1b.p4 pvs2b.p4

$ p4c --std p4-14 --target bmv2 --arch v1model pvs1.p4
$ p4c --std p4-14 --target bmv2 --arch v1model pvs2.p4
$ diff pvs1.json pvs2.json
```
The only difference in output was a couple of occurrences of the source file name in the debug info part of the BMv2 JSON files.  The 'functional part' was identical.

[pvs.zip](https://github.com/p4lang/p4c/files/4157159/pvs.zip)

**Comment by vgurevich** [comment]
I think that the main problem is that the keyword `mask` in P4_14 is overloaded with two different meanings:  sometimes it it's equivalent to `&&&` in P4_16 and sometimes it is equivalent to a bit-slice.

This is an unfortunate design choice that was made long time ago and this is a good example of this ambiguity. 

Given P4_16 flexibility,  BMv2 should be able to support such a construct.  

With regards to the support in the compiler it would be nice to have,  but if not,  then not.

**Comment by hesingh** [comment]
Let's first remove any ambiguity from p4-14 and p4-16 specs and then we can plan p4c code changes.  

There is this feedback from Antonin regarding p4-14.

https://github.com/p4lang/p4c/pull/2183#discussion_r374828806

**Comment by jafingerhut** [comment]
@vgurevich If you want support for it in p4c's P4_16 compiler, then it should be added to the language spec.

What is it intended to mean?  The P4_14 language spec already says that pvs (parser value set) entries are all ternary, with the control plane installing value/mask for each entry added to one.

Is P4_14 `my_pvs`, without the `mask 0xff00` supposed to mean an entry with value v1 and mask m1 masks a select key k if and only if (k & v1) == (v1 & m1)

then if you have `my_pvs`, it is supposed to mean an entry matches if and only if (k & v1 & 0xff00) == (v1 & m1 & 0xff00)?  That is, the extra mask in the source code is like a "global mask" on all entries of that parser value set?

**Comment by vgurevich** [comment]
@jafingerhut,

I do not think that there is a need to support this in P4_16. 

There might be some value in supporting this for P4_14. 

The semantics is that despite the fact that the original field is 16-bit-wide, the PVS can be 8- bit-wide and that only the middle bits of the field from the select() statement will be fed into the pvs. In P4_16 we can accomplish this with a bit- slice or an AND in the expression in the select() statement.

**Comment by hesingh** [comment]
If we support for p4-14, do we also support any translation to p4-16 or just skip any translation?

**Comment by vgurevich** [comment]
@hesing,

We do support P4_14 to P4_16 translation.  My point is that P4_16 already has the necessary  syntax to express this. Thus, it's a matter of P4_14 support only. Most targets should not need anything extra either.

**Comment by jafingerhut** [comment]
The punishment for asking for a feature is answering a bunch of questions about it from people who might implement and/or specify it in more detail :-)

Is this P4_14 program meaningful, and if so, what should it do?  Or is it an error?
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x0ff0 : next_parse_state_1;
        pvs1 mask 0xf0f0 : next_parse_state_2;
        default : ingress;
    }
}
```

**Comment by vgurevich** [comment]
@jafingerhut,

I'm traveling,  so it's a little difficult to type code on my phone. Please see my comment above that explains the semantics.  If that's not clear,  I'll provide P4_16 equivalent later today. The code makes sense in my opinion, although it's a spec extension and on the convoluted side

**Comment by jafingerhut** [comment]
Sure, no rush on this that I know of.  If a decision is made a month or three from now on whether the compiler is changed, or how, I don't think anyone is hanging on the edge of a cliff needing the answers soon.

**Comment by jafingerhut** [comment]
I will throw in one more P4_14 program snippet using this feature, so perhaps when answering you can also think about whether it should be an error, or if not, what the behavior should be:
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x0ff0 : next_parse_state_1;
        default : ingress;
    }
}

parser parser_inner_ethernet {
    extract(inner_ethernet);
    return select(latest.etherType) {
        pvs0 mask 0xfff0 : next_parse_state_1;   // note that the mask on pvs0 here is different than the one above
        default : ingress;
    }
}
```

**Comment by hesingh** [comment]
@jafingerhut Regarding "Note that v1.0.5 of the P4_14 language spec does not mention this syntax at all. In section 4.4 "BNF" v1.0.5 of the P4_14 language spec gives this grammar rule:"

I looked at the p4c code for p4-14 for the grammar.  See

https://github.com/p4lang/p4c/blob/master/frontends/parsers/v1/v1parser.ypp#L527

and this is why the `pvs0 mask 0xfff0` is accepted.   So, maybe the p4-14 spec needs an update.

**Comment by antoninbas** [comment]
@jafingerhut 

I actually did not realize that the P4_14 spec required implementations to support and individual mask for every PVS entry. If all PVS entries are supposed to be ternary (which is not supported in bmv2 - so I wonder if this was always the case, or was only introduced later on in a P4_14 spec revision), the ability to provide one "global" mask is definitely not as useful any more.

The example you give in https://github.com/p4lang/p4c/issues/2188#issuecomment-582631216 is interesting and definitely something I strived to accommodate in the bmv2 implementation. Not sure whether a high-speed ASIC can actually leverage the "sharing" of the PVS between parser states to save some memory. This is pretty much the only case that cannot be easily translated to P4_16 IMO. Another one may be using the same PVS (with or without mask) in 2 different parser states which use a different transition key (even if the bitwidths of the concatenated keys are the same) - see https://github.com/p4lang/behavioral-model/pull/859#issuecomment-581560962.

**Comment by vgurevich** [comment]
@jafingerhut ,

If I were to ascribe the semantics to the parser_value_set with mask, then I would use the mask to reduce the width of the pvs.

So, the very first example makes a lot of sense to me:
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x0ff0 : next_parse_state_1;
        default : ingress;
    }
}
```
recoded in P4_16 would look like this:
```
state parse_ethernet {
    pkt.extract(hdr.ethernet);
    transition select(hdr.ethernet.etherType[11:4]) {
        pvs0 : next_parse_state_1;
    }
}
```
This is a useful optimization, especially given the fact that P4_14 does not have bit slices.

However, the other examples do not make much sense: in the first case (with different masks) we can't have different slices within the same select() statement (in a good way) and in the second case (with masks 0x0FF0 and 0xFFF0) the actual requested width of the PVS will be different. 

Thus, `mask` is sometimes useful, but only when the actual mask is the same everywhere.

**Comment by jafingerhut** [comment]
@vgurevich Thanks for taking a close look and answering.  What you say makes sense to me.

So, for the examples where they do not make much sense, would you say it is reasonable that they make so little sense, that a P4_14 compiler could reasonably give error messages if you tried to compiler them?

**Comment by vgurevich** [comment]
@jafingerhut  I'd say that there is a choice: the compiler can issue an error or it can issue a warning, stating that it decided to ignore the mask.

**Comment by jafingerhut** [comment]
Sorry, the punishment continues :-)

How about this program?  Note there are no masks here, but the same pvs0 is used in one place with a select expression having a 16-bit wide value, and in a different parser state with a select expression having a 48-bit wide value:
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {    // <-- latest.etherType is a 16-bit field
        pvs0 : next_parse_state_1;
        default : ingress;
    }
}

parser parser_inner_ethernet {
    extract(inner_ethernet);
    return select(latest.dstAddr) {   // <-- latest.dstAddr is a 48-bit field
        pvs0 : next_parse_state_1;
        default : ingress;
    }
}
```
Would you say this should be:
(a) compile time error
(b) compile time warning
(c) legal

If (b) or (c), what should the width of pvs0 be?    If you say "48 bits wide", what if the control plane adds entries that are exact match in the upper 32 bits?  Should the first occurrence of pvs0 match anything ever?  If you say "16 bits wide", should the second occurrence of pvs0 match anything ever?

**Comment by vgurevich** [comment]
@jafingerhut Definitely a compile error.

**Comment by jafingerhut** [comment]
Sorry, yes, I should have reviewed the P4_14 language spec before asking, since it does mention this case already: "The width of the values is inferred from the place where the value set is referenced. If the set is used in multiple places and they would infer different widths, then the compiler must raise an error."

**Comment by jafingerhut** [comment]
OK, last one for tonight:
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x5555 : next_parse_state_1;
        default : ingress;
    }
}
```
Should that be legal, and case pvs0 to be an 8-bit wide value set, matching against only bit positions 0, 2, 4, 6, 8, 10, 12, and 14 of latest.etherType?

Or illegal, because there ought to be a restriction that the language only supports parser value set masks with consecutive 1 bits?

**Comment by vgurevich** [comment]
I think that it should be legal, but the usefulness will depend on the target. For example on Tofino it is not going to save you anything, so the compiler can ignore the mask when allocating the resources, but will communicate it to the driver/API.

**Comment by jafingerhut** [comment]
So same P4_14 source code, you would want the compiler to generate different control plane APIs when compiling to different targets?  That seems to be something we are trying to avoid in P4_16 at least with the P4Runtime API.

**Comment by vgurevich** [comment]
Not necessarily. In both cases the API to program the PVS will accept 16-bit values/masks.

You have more flexibility in P4_16. Consider these two constructs:
```
state parse_ethernet {
    pkt.extract(hdr.ethernet);
    transition select(hdr.ethernet.etherType[11:4]) {
        pvs0 : next_parse_state_1;
    }
}
```
and
```
state parse_ethernet {
    pkt.extract(hdr.ethernet);
    transition select(hdr.ethernet.etherType & 0x0ff0) {
        pvs0 : next_parse_state_1;
    }
}
```

The resulting code will be equivalent, but the APIs will be different: the first one will use 8-bit values/masks for the PVS, while the second will use 16-bit ones (but most probably store only 16 bits). 

Yet, it is a little confusing, but both approaches are useful. 

We do the same for the table keys (at least at Barefoot). Compare:
```
table t1 {
   key  = { h.ethernet.etherType[11:4] : exact @name{"ethertype_middle"};
   . . .
}
```
and 
```
table t1 {
   key  = { h.ethernet.etherType[11:4] & 0x0FF0: exact @name{"ethertype_middle"};
   . . .
}
```

**Comment by jafingerhut** [comment]
I lied about the earlier one being the last one for tonight :-)
P4_14 program snippet:
```
parser parse_ethernet {
    extract(ethernet);
    return select(latest.etherType) {
        pvs0 mask 0x0ff0 : next_parse_state_1;
        0xface : next_parse_state_2;
        default : ingress;
    }
}
```
P4_14 compile-time error, or legal?

If legal, how would you translate it to P4_16?

**Comment by vgurevich** [comment]
@jafingerhut -- probably a bug, because the masks are different. Or a warning that "mask" will be ignored, provided that the pvs is ternary.

As I probably alluded to before, this "mask" construct is only useful in very limited circumstances as a poor man's workaround for the lack of bit-slices in P4_14. Other than that there is little use for it and it is also confusing.

**Comment by vgurevich** [comment]
Actually, on the second thought, the statement can be compiled like this:
```
state parse_ethernet {
    pkt.extract(h.ethernet);
    transition select(h.ethernet.etherType & 0x0ff0) { // or use a bit slice
       pvs0 : next_parse_state_1;
      default: extra_state;
}

state extra_state {
    transition select(h.ethernet.etherType) {
        0xface: next_parse_state_2;
        default: accept;
    }
}
```

**Comment by hesingh** [comment]
The bmv2 backend does not support a ternary pvs.  I don't see the p4c frontend and translation code support any errors discussed here.   There is new code to be developed.  Should we write the new code or just stick to using p4-16 pvs?  Note, the v1 lang grammar already supports `pvs mask <hex>`.  If such p4-14 code is used, p4c will generate incorrect output.  If we don't write new code, we live with the incorrect output as a negative test case for p4c and p4-14.

**Comment by vgurevich** [comment]
@hesingh -- generally speaking the construct we are discussing is somewhat orthogonal to the match kind used by the  pvs.

Having said that, expanding BMv2 capabilities and also making parser value sets declarations cleaner and more specific (if not in P4_14, but in P4_16, that currently uses pragmas) would be a good thing to do.

**Comment by hesingh** [comment]
I think, p4-16 pvs is clean enough.  See this example.

`value_set<value_set_t>(4) pvs;`

The p4-14 pragma is gone and replaced by `(4)`.  p4-16 does not allow ternary or binary operations with pvs having tested the latest p4c code.  I don't think there is a need to allow.

There is one other piece of p4c code that needs to change if we clean up p4-14 and expand bmv2.  The translation has to provided the information to bmv2.  This is what my PR has developed the code for:  https://github.com/p4lang/p4c/pull/2194

**Comment by vgurevich** [comment]
@hesingh -- I think that there is another annotation to specify the match kind. 

I would've preferred:
```
value_set<value_set_t>(ternary, 4) pvs
```
but that requires a spec change.

**Comment by hesingh** [comment]
If @match annotation on a struct field is supported and the struct is used with pvs, would that work for you?  See

https://github.com/p4lang/behavioral-model/issues/857

**Comment by vgurevich** [comment]
That's precisely what I was referring to.

**Comment by jafingerhut** [comment]
Here is one alternative that seems fairly low effort, i.e. it doesn't have us considering two or three different ways that a P4_14 program might be auto-translated to P4_16, some of which might add extra states, and/or change the control plane API width of the parser value set:

+ Change p4c so that instead of its current behavior [1], issue a warning that p4c is ignoring the mask (with file and line number), and perhaps a link to a documentation page somewhere explaining the issue and alternatives.  If the developer is using p4c to auto-translate P4_14 to P4_16, they would still get that warning, and the linked docs would suggest ways that the developer could edit the resulting P4_16 program, if they wish.

[1] It seems that the current behavior of p4c is to silently accept P4_14 `pvs0 mask 0x0ff0 : next_parser_state;` and treat it identically to `pvs0 : next_parser_state` in every occurrence.

**Comment by hesingh** [comment]
The silent acceptance of p4c to accept p4-14 pvs with/without mask is incomplete compiler code.   It is incomplete code because one can see the p4-14 lang grammar accepts both types of pvs statements above but the rest of the p4c code does not cater to the differences.  

The code is completed in my PR since Antonin was on top of it for what p4-14 spec supports and I used his guidance to develop the code changes of the PR.  It is minor changes to the PR to catch error/warning conditions with p4-14 code for examples that Andy brought up in this PR.

**Comment by jafingerhut** [comment]
@hesingh Silent acceptance is bad, agreed, and I am not advocating for that, else I would not be participating in this discussion.

I suspect the code changes you have made so far are unlikely to cover all of the cases and examples discussed in comments here.  It seems to me like going to that work might be unnecessary and undesirable, if a warning is enough to point people at this feature.

From what I can tell about the answers, it seems that if one had ternary entries in the parser value set, not exact, then simply restricting the control plane to always using wildcard matches in the bit positions where a mask is 0, would implement the desired behavior.

Note I say it implements the desired behavior.  It does not implement one of the desired intents that it seems this feature was introduced for, which is to optimize space in a target used to store parser value set entries.  If we get correct behavior, though, not necessarily optimized for space, and issue a warning that it has not been optimized for space, we are not fooling anybody.

**Comment by jafingerhut** [comment]
Hmm, actually, said another way, maybe the code changes you have might actually do what I describe, except they do not issue a warning message, and to complete it would require BMv2 changes to support ternary parser value set entries.

**Comment by hesingh** [comment]
Yes, right now my PR simply detects p4-14 mask or not and sends the info to the backend.  The bmv2 backend has TODO code to treat the pvs entry as ternary and deal with (value, mask) pair. 

Regarding the warnings and errors,  I already said, it's pending code and the code changes are minor.

**Comment by mbudiu-vmw** [comment]
For now I will tag this as a bug; we will give an error if a value set is used with a mask.
We can improve on this later.
